precompiles/utils/src/precompile_set.rs

```diff
@@ -18,241 +18,389 @@
 	}
 }
 
-// INDIVIDUAL PRECOMPILE(SET)
+/// Precompiles are allowed to call this precompile.
+pub struct CallableByPrecompile<T = ForAllSelectors>(PhantomData<T>);
 
-/// A fragment of a PrecompileSet. Should be implemented as is it
-/// was a PrecompileSet containing only the precompile(set) it wraps.
-/// They can be combined into a real PrecompileSet using `PrecompileSetBuilder`.
-pub trait PrecompileSetFragment {
-	/// Instanciate the fragment.
-	fn new() -> Self;
+impl<T: SelectorFilter> PrecompileChecks for CallableByPrecompile<T> {
+	#[inline(always)]
+	fn callable_by_precompile(caller: H160, called_selector: Option<u32>) -> Option<bool> {
+		Some(T::is_allowed(caller, called_selector))
+	}
+}
 
-	/// Execute the fragment.
-	fn execute(&self, handle: &mut impl PrecompileHandle) -> Option<PrecompileResult>;
+fn is_address_eoa_or_precompile<R: pallet_evm::Config>(address: H160) -> bool {
+	let code_len = pallet_evm::AccountCodes::<R>::decode_len(address).unwrap_or(0);
 
-	/// Is the provided address a precompile in this fragment?
-	fn is_precompile(&self, address: H160) -> bool;
+	// 0 => either EOA or precompile without dummy code
+	if code_len == 0 {
+		return true;
+	}
 
-	/// Return the list of addresses covered by this fragment.
-	fn used_addresses(&self) -> Vec<H160>;
-}
+	// dummy code is 5 bytes long, so any other len means it is a contract.
+	if code_len != 5 {
+		return false;
+	}
 
-/// Wraps a stateless precompile: a type implementing the `Precompile` trait.
-/// Type parameters allow to define:
-/// - A: The address of the precompile
-/// - R: The recursion limit (defaults to 1)
-/// - D: If DELEGATECALL is supported (default to no)
-pub struct PrecompileAt<A, P, R = ForbidRecursion, D = ForbidDelegateCall> {
-	current_recursion_level: RefCell<u16>,
-	_phantom: PhantomData<(A, P, R, D)>,
+	// check code matches dummy code
+	let code = pallet_evm::AccountCodes::<R>::get(address);
+	&code == &[0x60, 0x00, 0x60, 0x00, 0xfd]
 }
 
-impl<A, P, R, D> PrecompileSetFragment for PrecompileAt<A, P, R, D>
-where
-	A: Get<H160>,
-	P: Precompile,
-	R: RecursionLimit,
-	D: DelegateCallSupport,
-{
-	#[inline(always)]
-	fn new() -> Self {
-		Self {
-			current_recursion_level: RefCell::new(0),
-			_phantom: PhantomData,
+/// Common checks for precompile and precompile sets.
+/// Don't contain recursion check as precompile sets have recursion check for each member.
+fn common_checks<R: pallet_evm::Config, C: PrecompileChecks>(
+	handle: &mut impl PrecompileHandle,
+) -> Result<(), fp_evm::PrecompileFailure> {
+	let code_address = handle.code_address();
+	let caller = handle.context().caller;
+
+	// Check DELEGATECALL config.
+	let allow_delegate_call = C::accept_delegate_call().unwrap_or(false);
+	if !allow_delegate_call && code_address != handle.context().address {
+		return Err(revert("Cannot be called with DELEGATECALL or CALLCODE"));
+	}
+
+	// Extract which selector is called.
+	let selector = handle.input().get(0..4).map(|bytes| {
```

Comment:
Does this need a bounds check (input size >= 4)?

Reply:
`get` returns `None` if the range is out of bound